sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,550 @@
+ * started if further updates are required.
+ *
+ * This call will also perform checks equivalent to an `epmp_check`
+ * call before modifying any registers.
+ *
+ * Updates will occur in the following sequence:
+ *
+ *   1. pmpaddr0-pmpaddr15
+ *   2. pmpcfg0-pmpcfg3
+ *
+ * @param count Number of entries that should have been configured.
+ * @returns `epmp_result_t`
+ */
+epmp_result_t epmp_transaction_end(epmp_entry_count_t count);
+
+/**
+ * EPMP configuration status codes.
+ */
+typedef enum {
+  kEpmpEntryConfigureOk = kEpmpOk,
+  kEpmpEntryConfigureError = kEpmpError,
+  kEpmpEntryConfigureBadArg = kEpmpBadArg,
+
+  /**
+   * Invalid addresses provided for the selected address mode.
+   */
+  kEpmpEntryConfigureBadRegion,
+
+  /**
+   * The transaction is in a bad state. This can be caused by:
+   *  * Missing call the `epmp_transaction_start`.
+   *  * An attempt to configure more entries than was initially specified.
+   *  * A prior configuration attempt encountered an error.
+   */
+  kEpmpEntryConfigureBadTransaction,
+
+  /**
+   * The requested entry is out of range or has previously been configured
+   * in this transaction. Each entry may only be configured once per
+   * transaction.
+   */
+  kEpmpEntryConfigureBadEntry,
+
+  /**
+   * Encoding the entry would interfere with a different pre-existing entry.
+   *
+   * New entries will be rejected if they:
+   *  * Modify the start or end address of an adjacent TOR entry.
+   *  * Would result in an address being used in both a NAPOT/NA4 entry and a
+   *    TOR entry.
+   */
+  kEpmpEntryConfigureConflict,
+} epmp_entry_configure_result_t;
+
+/**
+ * Disable EPMP address matching.
+ *
+ * Address matching is disabled for `entry`. The pmpaddr for `entry` is set
+ * to the value of `region.start` which must also match `region.end` (i.e.
+ * the length of the region should be 0).
+ *
+ * IMPORTANT: the entry will not be updated until `epmp_transaction_end`
+ * is called.
+ *
+ * Example:
+ *
+ *   ...
+ *   res0 = epmp_transaction_start(kEpmpEntryCount2);
+ *   res1 = epmp_entry_configure_off(kEpmpEntry0,
+ *              (epmp_region_t){0},
+ *              kEpmpPermUnlocked);
+ *   res2 = epmp_entry_configure_off(kEpmpEntry1,
+ *              (epmp_region_t){ .start = 0x10, .end = 0x10 },
+ *              kEpmpPermLockedNoAccess);
+ *   res3 = epmp_transaction_end(kEpmpEntryCount2);
+ *   ...
+ *
+ * Result:
+ *
+ *   Entry | Value of `pmpaddr` | Value of `pmpcfg` |
+ *   ======+====================+===================+
+ *       0 |   0x00 (0x00 >> 2) |         0b0000000 |
+ *       1 |   0x04 (0x10 >> 2) |         0b1000000 |
+ *
+ * @param entry Entry index to update (0 <= `entry` < `kEpmpNumRegions`)
+ * @param region Updated value for pmpaddr (see above).
+ * @param permissions Updated permissions to write to pmpcfg for `entry`.
+ * @return `epmp_entry_configure_result_t`.
+ */
+epmp_entry_configure_result_t epmp_entry_configure_off(epmp_entry_t entry,
```

Comment:
I mostly added it for flexibility - it gives the user a way to set the `pmpaddr` and `pmpcfg` registers for the disabled entry to the values they want. For example, if a boot stage reserves the first 4 entries and then only needs to use 3 they might want to lock the fourth entry even though it is OFF to prevent someone inserting an extra high priority rule there (assuming `RLB = 0` by then of course).

If the entry being disabled is used as a TOR start address it also forces the user to put in that same address when they disable it. I have mixed feelings about whether that is good or not.

Reply:
I think either way is probably fine.

Reply:
I think I'll leave it as-is for now. We can probably change this later if we find it not to be useful.